File: July/Day3/find_the_kth_character_in_string_game_i.py
#TC: 2^t , t is the power 1, 2, 3,

class Solution:
    def kthCharacter(self, k: int) -> str:
        prevString = 'a'
        while len(prevString) < k:
            size = len(prevString)
            currString =''
            for ch in prevString:
                nextChar = chr(ord('a') + ((ord(ch) - ord('a') + 1) % 26))
                currString+=nextChar
            prevString+= currString
        return prevString[k-1] 


# Logarithmic approach: O(log(k))

class Solution:
    def kthCharacter(self, k: int) -> str:
        ans = 0
        while k != 1:
            power = k.bit_length() - 1
            if (1 << power) == k:
                power -= 1
            k -= 1 << power
            ans += 1
        return chr(ord("a") + ans)



# Bit Count: TC (O(1))
class Solution:
    def kthCharacter(self, k: int) -> str:
        return chr(ord("a")+ (k-1).bit_count())

File: July/Day3/find_the_kth_character_in_string_game_i.cc
class Solution {
public:
    char kthCharacter(int k) {
        std::string prevString = "a";
        while (prevString.length() < k) {
            std::string currString = "";
            for (char ch : prevString) {
                char nextChar = 'a' + ((ch - 'a' + 1) % 26);
                currString += nextChar;
            }
            prevString += currString;
        }
        return prevString[k - 1];
    }
};


// Logarithmic approach: O(log(k))

class Solution {
public:
    char kthCharacter(int k) {
        int ans = 0;
        while (k != 1) {
            int power = 31 - __builtin_clz(k); // bit_length - 1
            if ((1 << power) == k) power--;
            k -= 1 << power;
            ans++;
        }
        return 'a' + ans;
    }
};


// Bit Count (O(1))
class Solution {
public:
    char kthCharacter(int k) {
        int bitCount = __builtin_popcount(k - 1);
        return 'a' + bitCount;
    }
};


